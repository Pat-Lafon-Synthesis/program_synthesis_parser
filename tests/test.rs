use std::{fs::File, io::Read};

use program_synthesis_parser;

macro_rules! make_test {
    ($test_name:tt) => {
        #[test]
        fn $test_name() {
            let parser = program_synthesis_parser::spec::ProblemParser::new();
            let mut file =
                File::open(format!("tests/benchmarks/{}.mls", stringify!($test_name))).unwrap();
            let mut buffer = String::new();
            file.read_to_string(&mut buffer).unwrap();
            insta::assert_debug_snapshot!(parser.parse(&buffer).unwrap());
        }
    };
}

make_test!(bool_always_false);
make_test!(bool_always_true);
make_test!(bool_band);
make_test!(bool_bor);
make_test!(bool_impl);
make_test!(bool_neg);


make_test!(expr_boolean);
make_test!(expr_div);
make_test!(expr_sub);
make_test!(expr);


make_test!(list_append);
make_test!(list_compress);
make_test!(list_concat);
make_test!(list_drop);
make_test!(list_dropeven);
make_test!(list_even_parity);
make_test!(list_filter);
make_test!(list_fold);
make_test!(list_hd);
make_test!(list_inc);
make_test!(list_last);
make_test!(list_last2);
make_test!(list_length);
make_test!(list_make);
make_test!(list_map);
make_test!(list_nth);
make_test!(list_pairwise_swap);
make_test!(list_range);
make_test!(list_rev_append);
make_test!(list_rev_fold);
make_test!(list_rev_snoc);
make_test!(list_rev_tailcall);
make_test!(list_snoc);
make_test!(list_sort_sorted_insert);
make_test!(list_sorted_insert);
make_test!(list_stutter);
make_test!(list_sum);
make_test!(list_take);
make_test!(list_tl);


make_test!(nat_add);
make_test!(nat_iseven);
make_test!(nat_max);
make_test!(nat_mul);
make_test!(nat_pred);
make_test!(nat_sub);


make_test!(tree_balanced);
make_test!(tree_binsert);
make_test!(tree_collect_leaves);
make_test!(tree_count_leaves);
make_test!(tree_count_nodes);
make_test!(tree_height);
make_test!(tree_inorder);
make_test!(tree_lastleft);
make_test!(tree_map);
make_test!(tree_nodes_at_level);
make_test!(tree_notexist);
make_test!(tree_postorder);
make_test!(tree_preorder);
make_test!(tree_sum);