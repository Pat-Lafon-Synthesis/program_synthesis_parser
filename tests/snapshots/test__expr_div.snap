---
source: tests/test.rs
expression: parser.parse(&buffer).unwrap()
---
SynthProblem {
    imports: [],
    decls: [
        TypeDecl(
            "nat",
            Type(
                Variant([("O", Type(Tuple([]))), ("S", Type(Named("nat")))]),
            ),
        ),
        TypeDecl(
            "exp",
            Type(
                Variant([("INT", Type(Named("nat"))), ("ADD", Type(Tuple([Type(Named("exp")), Type(Named("exp"))]))), ("SUB", Type(Tuple([Type(Named("exp")), Type(Named("exp"))]))), ("MUL", Type(Tuple([Type(Named("exp")), Type(Named("exp"))]))), ("DIV", Type(Tuple([Type(Named("exp")), Type(Named("exp"))])))]),
            ),
        ),
        TypeDecl(
            "cmp",
            Type(
                Variant([("LT", Type(Tuple([]))), ("EQ", Type(Tuple([]))), ("GT", Type(Tuple([])))]),
            ),
        ),
        ExpDecl(
            "compare",
            Exp(
                Fix("compare", Type(Arrow(Type(Arrow(Type(Named("nat")), Type(Named("nat")))), Type(Named("cmp")))), Exp(Func(Param { name: "x1", ty: Type(Named("nat")) }, Exp(Func(Param { name: "x2", ty: Type(Named("nat")) }, Exp(Match(Exp(Var("x1")), [(Ctor("O", Wildcard), Exp(Match(Exp(Var("x2")), [(Ctor("O", Wildcard), Exp(CTor("EQ", Exp(Tuple([]))))), (Ctor("S", Wildcard), Exp(CTor("LT", Exp(Tuple([])))))]))), (Ctor("S", Var("x1p")), Exp(Match(Exp(Var("x2")), [(Ctor("O", Wildcard), Exp(CTor("GT", Exp(Tuple([]))))), (Ctor("S", Var("x2p")), Exp(App(Exp(App(Exp(Var("compare")), Exp(Var("x1p")))), Exp(Var("x2p")))))])))]))))))),
            ),
        ),
        ExpDecl(
            "add",
            Exp(
                Fix("add", Type(Arrow(Type(Arrow(Type(Named("nat")), Type(Named("nat")))), Type(Named("nat")))), Exp(Func(Param { name: "n1", ty: Type(Named("nat")) }, Exp(Func(Param { name: "n2", ty: Type(Named("nat")) }, Exp(Match(Exp(Var("n1")), [(Ctor("O", Wildcard), Exp(Var("n2"))), (Ctor("S", Var("n1p")), Exp(CTor("S", Exp(Tuple([Exp(App(Exp(App(Exp(Var("add")), Exp(Var("n1p")))), Exp(Var("n2"))))])))))]))))))),
            ),
        ),
        ExpDecl(
            "sub",
            Exp(
                Fix("sub", Type(Arrow(Type(Arrow(Type(Named("nat")), Type(Named("nat")))), Type(Named("nat")))), Exp(Func(Param { name: "n1", ty: Type(Named("nat")) }, Exp(Func(Param { name: "n2", ty: Type(Named("nat")) }, Exp(Match(Exp(Var("n1")), [(Ctor("O", Wildcard), Exp(CTor("O", Exp(Tuple([]))))), (Ctor("S", Var("n1p")), Exp(Match(Exp(Var("n2")), [(Ctor("O", Wildcard), Exp(Var("n1"))), (Ctor("S", Var("n2p")), Exp(App(Exp(App(Exp(Var("sub")), Exp(Var("n1p")))), Exp(Var("n2p")))))])))]))))))),
            ),
        ),
        ExpDecl(
            "mul",
            Exp(
                Fix("mul", Type(Arrow(Type(Arrow(Type(Named("nat")), Type(Named("nat")))), Type(Named("nat")))), Exp(Func(Param { name: "n1", ty: Type(Named("nat")) }, Exp(Func(Param { name: "n2", ty: Type(Named("nat")) }, Exp(Match(Exp(Var("n1")), [(Ctor("O", Wildcard), Exp(CTor("O", Exp(Tuple([]))))), (Ctor("S", Var("n1p")), Exp(App(Exp(App(Exp(Var("add")), Exp(Var("n2")))), Exp(App(Exp(App(Exp(Var("mul")), Exp(Var("n1p")))), Exp(Var("n2")))))))]))))))),
            ),
        ),
        ExpDecl(
            "div",
            Exp(
                Fix("div", Type(Arrow(Type(Arrow(Type(Named("nat")), Type(Named("nat")))), Type(Named("nat")))), Exp(Func(Param { name: "n1", ty: Type(Named("nat")) }, Exp(Func(Param { name: "n2", ty: Type(Named("nat")) }, Exp(Match(Exp(Var("n2")), [(Ctor("O", Wildcard), Exp(CTor("O", Exp(Tuple([]))))), (Ctor("S", Var("n2p")), Exp(Match(Exp(Var("n1")), [(Ctor("O", Wildcard), Exp(CTor("O", Exp(Tuple([]))))), (Ctor("S", Var("n1p")), Exp(Match(Exp(App(Exp(App(Exp(Var("compare")), Exp(Var("n1")))), Exp(Var("n2")))), [(Ctor("LT", Wildcard), Exp(CTor("O", Exp(Tuple([]))))), (Ctor("EQ", Wildcard), Exp(CTor("S", Exp(Tuple([Exp(CTor("O", Exp(Tuple([]))))]))))), (Ctor("GT", Wildcard), Exp(CTor("S", Exp(Tuple([Exp(App(Exp(App(Exp(Var("div")), Exp(App(Exp(App(Exp(Var("sub")), Exp(Var("n1")))), Exp(Var("n2")))))), Exp(Var("n2"))))])))))])))])))]))))))),
            ),
        ),
    ],
    synth_type: Type(
        Arrow(Type(Named("exp")), Type(Named("nat"))),
    ),
    spec: UIOEs(
        [
            Example {
                input: [
                    Exp(
                        CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))]))),
                    ),
                ],
                output: Exp(
                    CTor("S", Exp(CTor("O", Exp(Tuple([]))))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("ADD", Exp(Tuple([Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))])))), Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))))))))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("MUL", Exp(Tuple([Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))])))), Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))))))))))))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("MUL", Exp(Tuple([Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))])))), Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))))))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("SUB", Exp(Tuple([Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))))])))), Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("S", Exp(CTor("O", Exp(Tuple([]))))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("SUB", Exp(Tuple([Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))))))])))), Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("DIV", Exp(Tuple([Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))))])))), Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("DIV", Exp(Tuple([Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))))))])))), Exp(CTor("INT", Exp(Tuple([Exp(CTor("S", Exp(CTor("S", Exp(CTor("S", Exp(CTor("O", Exp(Tuple([]))))))))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("S", Exp(CTor("O", Exp(Tuple([]))))),
                ),
            },
        ],
    ),
}
